File: 백준/Silver/3184. 양/양.kt
package org.example

val dx = intArrayOf(-1, 1, 0, 0)
val dy = intArrayOf(0, 0, -1, 1)

fun main() {
    val br = System.`in`.bufferedReader()
    val (r, c) = br.readLine().split(" ").map { it.toInt() }
    val graph = Array(r) { br.readLine().toCharArray() }

    var totalSheep = 0
    var totalWolf = 0

    for (i in 0 until r) {
        for (j in 0 until c) {
            if (graph[i][j] == 'v' || graph[i][j] == 'o') {
                val (sheep, wolf) = dfs(graph, i, j)
                if (sheep > wolf) {
                    totalSheep += sheep
                } else {
                    totalWolf += wolf
                }
            }

        }
    }
    println("$totalSheep $totalWolf")
    br.close()
}

fun dfs(graph: Array<CharArray>, x: Int, y: Int): Pair<Int, Int> {
    if (x !in graph.indices || y !in graph[0].indices || graph[x][y] == '#') {
        return Pair(0, 0)
    }

    var sheep = 0
    var wolf = 0

    if (graph[x][y] == 'v') wolf++
    else if (graph[x][y] == 'o') sheep++

    graph[x][y] = '#'

    for (i in 0..3) {
        val (ns, nw) = dfs(graph, x + dx[i], y + dy[i])
        sheep += ns
        wolf += nw
    }
    return Pair(sheep, wolf)
}

File: 백준/Silver/3184. 양/양.py
import sys
from collections import deque

input = sys.stdin.readline

directions = [
    (-1, 0),
    (1, 0),
    (0, -1),
    (0, 1)
]


def bfs(graph, start_x, start_y):
    sheep = 0
    wolf = 0
    queue = deque([(start_x, start_y)])
    if graph[start_x][start_y] == 'v':
        # 늑대
        wolf += 1
    elif graph[start_x][start_y] == 'o':
        # 양
        sheep += 1
    graph[start_x][start_y] = '#'

    while queue:
        cx, cy = queue.popleft()

        for (dx, dy) in directions:
            nx = cx + dx
            ny = cy + dy
            if 0 <= nx < len(graph) and 0 <= ny < len(graph[0]) and graph[nx][ny] != '#':
                queue.append((nx, ny))
                if graph[nx][ny] == 'v':
                    # 늑대
                    wolf += 1
                elif graph[nx][ny] == 'o':
                    # 양
                    sheep += 1
                graph[nx][ny] = '#'
            elif nx < 0 or nx >= len(graph) or ny < 0 or ny >= len(graph[0]):
                return 0, 0
    if wolf >= sheep:
        return 0, wolf
    elif wolf < sheep:
        return sheep, 0


r, c = map(int, input().split())  # 행 열

graph = [list(map(str, input().strip())) for _ in range(r)]

total_wolf = 0
total_sheep = 0
for i in range(r):
    for j in range(c):
        if graph[i][j] == 'o' or graph[i][j] == 'v':
            sheep, wolf = bfs(graph, i, j)
            total_wolf += wolf
            total_sheep += sheep
print(total_sheep, total_wolf)
